### ISSUE TITLE
rethinkdb 2.0.4 memory usage increases over time 

### ISSUE THREAD

**Issue by balasuar** [rethinkdb/rethinkdb#4663]
When running RethinkDB on our EC2 instance, memory use increases over time. This occurs on every machine. Our usage is very simple, we store a number of JSON documents that are used to generate content pages which are served via HTML.

We utilize both browser and NGINX caching, so the actual hits to the underly rethinkdb on a per request basis are (in theory) minimal.

Here is an image that shows the memory increase, and the effect of restarting rethink.

![2015-08-11 at 11 33 am](https://cloud.githubusercontent.com/assets/278048/9206810/dbf8b1b6-401e-11e5-8378-478b6df9e312.png)

Output of the db status call: https://gist.github.com/balasuar/cb555b4a1cac865e2a4e

For completeness sake, here is the code that interfaces with rethinkdb:

https://gist.github.com/balasuar/5ee27d7a6e0de8914fe2

```
rethinkdb 2.0.4~0trusty (GCC 4.8.2)
Linux ip-XXXX-generic #63-Ubuntu SMP Wed Sep 3 21:30:07 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux
EC2 t2.medium instance
4 gb ram
rethink dash: 1.16.4
```

**Comment by danielmewes** [comment]
Thanks for the issue report @balasuar . This looks pretty bad. We will look into this.

@balasuar Was the status print from shortly after a restart? If you get a chance, could you do another status table dump when the memory consumption is high?

**Comment by danielmewes** [comment]
I assume the memory usage plot is for a single server?

Do you know how it's measured? Does it show virtual or actually paged-in memory (sometimes called RES)? Is this only of the `rethinkdb` process or for the whole system?

**Comment by balasuar** [comment]
@danielmewes The status dump is from after a restart. I will add a status dump after memory gets high enough again.

The memory usage plot is for a single server, and for only the rethinkdb process(es).

According to @newrelic docs it looks like it paged-in memory (similar to `free`). https://docs.newrelic.com/docs/servers/new-relic-servers-linux/troubleshooting/incorrect-memory-usage-metrics-individual-processes-servers-linux

**Comment by danielmewes** [comment]
Thanks for the information @balasuar . I don't quite understand what New Relic is measuring from the page you linked to. It's talking about "physical memory" first, but then also says "It uses the reported virtual set size for a process and multiplies that by that process's number of active instances.". So I'm not sure if the reported value is physical memory or virtual memory.

Just to make sure, would you mind also checking memory usage with a tool such as `top` or `htop` the next time the server is using a lot of RAM? The relevant column in these tools is "RES".

**Comment by balasuar** [comment]
@danielmewes I'll attach output from `top` as soon as I can. FWIW, I have checked top myself during high memory situations, and it shows the amount of free memory consistent with what newrelic reports.

**Comment by balasuar** [comment]
Currently at 1.3GB of Memory.

Here is what is being reported by `top.`
```
top - 05:01:29 up 7 days, 12:32,  1 user,  load average: 0.02, 0.02, 0.05
Tasks: 118 total,   1 running, 117 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0.0 us,  0.0 sy,  0.0 ni,100.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem:   4046804 total,  3169488 used,   877316 free,   188832 buffers
KiB Swap:        0 total,        0 used,        0 free.   791248 cached Mem

USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
ubuntu    20   0  353100 105864  14112 S   0.0  2.6   1:52.84 rethinkdb
ubuntu    20   0  172344  11560    216 S   0.0  0.3   0:00.00 rethinkdb
ubuntu    20   0  956700 201200   6160 S   0.0  5.0   0:22.24 rethinkdb
ubuntu    20   0 1368344 603328   4740 S   0.0 14.9   1:15.77 rethinkdb
ubuntu    20   0 1303836 529128   6144 S   0.0 13.1   0:53.32 rethinkdb
ubuntu    20   0 1259284 344464   6736 S   0.0  8.5   0:09.91 node
ubuntu    20   0  939648 220444   5884 S   0.0  5.4   0:04.10 node
root      20   0  105628   4208   3224 S   0.0  0.1   0:00.00 sshd
ubuntu    20   0  105628   1908    920 S   0.0  0.0   0:00.05 sshd
ubuntu    20   0   21372   3808   1752 S   0.0  0.1   0:00.05 bash
ubuntu    20   0   23680   1624   1116 R   0.0  0.0   0:00.37 top
```

**Comment by danielmewes** [comment]
Thanks for the top output @balasuar. That's quite helpful. 
The interesting thing is that the memory usage is spread over multiple rethinkdb processes. Usually RethinkDB uses only one process (plus a second one which should use almost no memory though). There is one exception to this, in that we fork additional processes when executing `r.js` or `r.http` terms. 

I assume you're running only one instance of RethinkDB on this server? Or are you running multiple ones on the same machine?

Are you using r.js in your application?
if the answer to that is yes, I strongly suspect that our memory leak lies in that part of our code, which would be a great start for us to isolate the problem.

**Comment by balasuar** [comment]
@danielmewes Correct, we only start with one instance of Rethink. It's managed by supervisor.

The output of `ps` shows that supervisor started one instance of rethink, which spawned another. That second instance spawned three more.

```
ubuntu@ip-172-31-16-26:~$ ps -ef | grep rethink
ubuntu   10163   234  0 Aug11 ?        00:01:29 rethinkdb --bind all
ubuntu   10165 10163  0 Aug11 ?        00:00:00 rethinkdb --bind all
ubuntu   10239 10165  0 Aug11 ?        00:00:24 rethinkdb --bind all
ubuntu   10328 10165  0 Aug11 ?        00:00:54 rethinkdb --bind all
ubuntu   10329 10165  0 Aug11 ?        00:00:43 rethinkdb --bind all
ubuntu   13116 13097  0 19:09 pts/0    00:00:00 grep --color=auto rethink
ubuntu@ip-172-31-16-26:~$ ps -ef | grep supervisord
root       234     1  0 Aug06 ?        00:01:56 /usr/bin/python /usr/bin/supervisord --nodaemon --configuration /etc/supervisor/supervisord.conf
ubuntu   13120 13097  0 19:10 pts/0    00:00:00 grep --color=auto supervisord
ubuntu@ip-172-31-16-26:~$
```

**Comment by balasuar** [comment]
@danielmewes 
> Are you using r.js in your application?

We are using r.js for two queries.

**Comment by danielmewes** [comment]
Awesome, that information is very useful.
This gives us a strong indication that the memory leak is in our `r.js()` implementation.
Would you be able to share the two `r.js` queries you mentioned?

**Comment by balasuar** [comment]
Yes. I am afk at the moment, however in my initial comment I have a link to a gist that has all our js code. 

-Alan
➡⬇↘👊🏾

**Comment by danielmewes** [comment]
Ah, I forgot about that, thanks! Will see if I can reproduce this.

**Comment by danielmewes** [comment]
I was able to reproduce the problem by running one of your `r.js` queries repeatedly.

The JS worker processes initially used about 40-50 MB of RAM, but use more and more over time (I got them to ~270 MB each within half an hour, and there are 8 of them).

The next step is to figure out what they are using that memory for, and why it's growing.

In the worst case, we could probably restart the JS worker processes periodically to clean up their memory. It would be nicer though to fix this properly.

**Comment by danielmewes** [comment]
This indeed turned out to be a memory leak in the JS worker code. We missed to explicitly free some V8 values.
Now each worker process stays at ~25 MB

A fix is in branch daniel_4663, and is in code review 3154 by @Tryneus.

**Comment by danielmewes** [comment]
@balasuar We have released RethinkDB 2.1.2 with a fix for this. We're also going to release a 2.0 point release (2.0.5) next week, in case you want to stay on the 2.0 branch.

**Comment by balasuar** [comment]
Great. I'll migrate to 2.1.2!

Thanks for addressing this fix so quickly.

**Comment by balasuar** [comment]
Here is a side by side comparison a machine that was upgraded from 2.0.4 to 2.1.2, and a machine still running 2.0.4

![2015-09-01 at 10 17 am](https://cloud.githubusercontent.com/assets/278048/9611174/1537c640-5093-11e5-8331-08258f02761d.png)

![2015-09-01 at 10 19 am](https://cloud.githubusercontent.com/assets/278048/9611180/1b1ef362-5093-11e5-9f98-769f23ac39e9.png)

**Comment by danielmewes** [comment]
@balasuar Happy to see the fix is working. Thanks for sharing the plots.
We've also released RethinkDB 2.0.5 yesterday, which you can use as a fully compatible in-place replacement for your 2.0.4 box. It doesn't require any migration, and also has the fix for this memory leak.
